bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/io/RefreshingOAuth2CredentialsInterceptor.java

```diff
@@ -76,339 +76,296 @@
+      refreshing = true;
+
+      try {
+        this.futureToken = executor.submit(new Callable<HeaderCacheElement>() {
+          @Override
+          public HeaderCacheElement call() throws Exception {
+            HeaderCacheElement newToken = refreshCredentials();
+            return updateToken(newToken);
+          }
+        });
+      } catch (RuntimeException e) {
+        refreshing = false;
+        throw e;
       }
+
+      return this.futureToken;
     }
-    return true;
   }
 
-  /**
-   * Calls {@link com.google.auth.oauth2.OAuth2Credentials#refreshAccessToken()}. In case of an
-   * IOException, retry the call as per the {@link com.google.api.client.util.BackOff} policy
-   * defined by {@link com.google.cloud.bigtable.config.RetryOptions#createBackoff()}.
-   *
-   * <p>This method retries until one of the following conditions occurs:
-   *
-   * <ol>
-   * <li>An OAuth request was completed. If the value is null, return an exception.
-   * <li>A non-IOException Exception is thrown - return an error status
-   * <li>All retries have been exhausted, i.e. when the Backoff.nextBackOffMillis() returns
-   *     BackOff.STOP
-   * <li>An interrupt occurs.
-   * </ol>
-   *
-   * @return HeaderCacheElement containing either a valid {@link com.google.auth.oauth2.AccessToken}
-   *     or an exception.
-   */
-  protected HeaderCacheElement refreshCredentialsWithRetry() {
-    final BackOff backoff = retryOptions.createBackoff();
-
-    while (true) {
+  private HeaderCacheElement updateToken(HeaderCacheElement newToken) {
+    synchronized (lock) {
       try {
-        logger.info("Refreshing the OAuth token");
-        AccessToken newToken = credentials.refreshAccessToken();
-        return new HeaderCacheElement(newToken);
-      } catch (IOException exception) {
-        logger.warn("Got an unexpected IOException when refreshing google credentials.", exception);
-        // An IOException occurred. Retry with backoff.
-        // Given the backoff, either sleep for a short duration, or terminate if the backoff has
-        // reached its configured timeout limit.
-        RetryState retryState = getRetryState(backoff);
-        if (retryState != RetryState.PerformRetry) {
-          return new HeaderCacheElement(exception);
-        } // else Retry.
-
-      } catch (Exception e) {
-        logger.warn("Got an unexpected exception while trying to refresh google credentials.", e);
-        return new HeaderCacheElement(new IOException("Could not read headers", e));
+        // Update the token only if the new token is good or the old token is bad
+        CacheState newState = newToken.getCacheState();
+        boolean newTokenOk = newState == CacheState.Good || newState == CacheState.Stale;
+        CacheState oldCacheState = headerCache.getCacheState();
+        boolean oldTokenOk = oldCacheState == CacheState.Good || oldCacheState == CacheState.Stale;
+
+        if (newTokenOk || !oldTokenOk) {
```

Comment:
A new good or stale token will always overwrite the old token. This is mainly to prevent a new pre-emptive bad token from expiring a stale old token.